cpp/include/raft/core/mdspan.hpp

```diff
@@ -304,4 +304,24 @@ RAFT_INLINE_FUNCTION auto unravel_index(Idx idx,
   }
 }
 
+/**
+ * @brief Create a copy of the given mdspan with const element type
+ * @tparam mdspan_type Expected type raft::host_mdspan or raft::device_mdspan
+ * @param mds raft::host_mdspan or raft::device_mdspan object
+ * @return raft::host_mdspan or raft::device_mdspan with vector_extent
+ *         depending on AccessoryPolicy
+ */
+template <typename mdspan_type, typename = enable_if_mdspan<mdspan_type>>
+auto make_const_mdspan(mdspan_type mds)
+{
+  using const_element_t = std::add_const_t<typename mdspan_type::element_type>;
+  using const_accessor_t =
```

Comment:
The definition of `const_accessor_t` unexpectedly type-erases the user's accessor type.  This could be surprising to users, and it could also prevent valid code from building.  For example, users might write their own accessor `my_custom_accessor`, and a function `device_function` that takes it.

```c++
template<class T>
void device_function(mdspan<const T, dextents<size_t, 2>, layout_left, my_custom_accessor<const T>> m);
```

`make_const_mdspan` would unexpectedly change the accessor.  This would break stdblas, for example, since it uses custom accessors.  Users might also be surprised that their `device_accessor` gets type-erased into `host_device_accessor`.

The fix is to add a customization point `accessor_of_const`, which defines how to get an accessor that is the "const version" of a given accessor.

```c++
template<class ElementType>
default_accessor<std::add_const_t<ElementType>>
accessor_of_const(default_accessor<ElementType> a)
{
  return {a};
}

template<class ElementType>
host_accessor<default_accessor<std::add_const_t<ElementType>>>
accessor_of_const(host_accessor<default_accessor<ElementType>> a)
{
  return {a};
}

template<class ElementType>
device_accessor<default_accessor<std::add_const_t<ElementType>>>
accessor_of_const(device_accessor<default_accessor<ElementType>> a)
{
  return {a};
}

template<class ElementType>
host_device_accessor<default_accessor<std::add_const_t<ElementType>>>
accessor_of_const(host_device_accessor<default_accessor<ElementType>> a)
{
  return {a};
}
```

Users could specialize this for their own accessor types.  Then, `make_const_mdspan` would be defined like this.

```
template<class ElementType, class Extents, class Layout, class Accessor>
auto make_const_mdspan(mdspan<ElementType, Extents, Layout, Accessor> mds)
{
  auto acc_c = accessor_of_const(mds.accessor()); 
  return mdspan<std::add_const_t<ElementType>, Extents, Layout, decltype(acc_c)>{
    mds.data_handle(),
    mds.mapping(),
    acc_c};
}
```

Reply:
If I remember right, `host_device_accessor` is just an alias to some type that has `bool` template parameters for "host accessible" and "device accessible"; if so, that would make the specialization logic simpler.

The point is that you only specialize `accessor_of_const` for accessor types where you know how to get the accessor-of-const.  Not every accessor type is required to have a natural accessor-of-const conversion.  If it does, it may not necessarily be exposed as a converting constructor.  (Think of accessors that view open files or remote PGAS memory, for example.)